File: 2022/go/07.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"strconv"
)

func main() {
	file, _ := os.Open("../data/07-1.txt")
	sc := bufio.NewScanner(file)

	var currDir = []string{"root"}
	var uniq = make(map[string]int)

	for sc.Scan() {
		if strings.Contains(sc.Text(), "$ cd") {
			currDir = append(currDir, strings.Split(sc.Text(), " ")[2])

			if strings.Contains(sc.Text(), "..") {
				currDir = currDir[:len(currDir)-2]
			} else {
				uniq[strings.Join(currDir, ",")] = 0
			}
		}
		if !strings.Contains("dir", sc.Text()) && !strings.Contains("ls", sc.Text()) {
			var fileSize, _ = strconv.Atoi(strings.Split(sc.Text(), " ")[0])
			for i := 1; i <= len(currDir); i++ {
				uniq[strings.Join(currDir[:i], ",")] += fileSize
			}
		}
	}
	var ans1 = 0
	for _, v := range uniq {
		if v <= 100000 {
			ans1 += v
		}
		if v >= 30000000 && (70000000 - uniq["root"]) <= v {
			// part 2 answer
			fmt.Println(v)
		}
	}
	// part 1 answer
	fmt.Println(ans1)
}

File: 2022/python/07.py
text = open("../data/07-1.txt").read().split("\n")

# simulate a directory structure with a list
currDir = ["root"]
dirsDict = {currDir[0]: 0}
for line in text:
  if "$ cd " in line:
    if ".." in line:
      currDir = currDir[:-1]
    elif "/" not in line:
      # this is the only condition where we add a directory
      currDir.append(line.split(" ")[-1])
      dirsDict["/".join(currDir)] = 0
  elif "dir " not in line and "$ ls" not in line:
    # when you add a filesize, you have to do it recursively
    for i in range(1, len(currDir)+1):
      dirsDict["/".join(currDir[:i])] += int(line.split(" ")[0])

# part 1 answer
print(sum([v for v in dirsDict.values() if v <= 100000]))
# part 2 answer
print(min([v for v in dirsDict.values() if v >= 30000000 - (70000000 - dirsDict["root"])]))